### ISSUE TITLE
Need to understand how INCLUDE CURVES works

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#1290]
I'm a little confused by the behaviour of INCLUDE_CURVES and hope that maybe once I understand it I can implement in a consistent manner how 2D objects are loaded into the scene.

Below are all the scenarios I can think of. A few things which I'm confused by:

 1. Isn't it extra work for the iterator to process it twice in Scenario A? Shouldn't it pick one and stick to it?
 2. In Scenario B how does it decide which representation to process? What if there were more than one 2D representation?
 3. I guess scenario C has a similar priority but for non 2D shapes?

In the past, I generally disabled INCLUDE_CURVES by default because I noticed some elements that had 3D bodies would be processed as 2D curves instead. But this was a long time ago and maybe I wasn't using the iterator, or maybe there is something I haven't discovered yet with the iterator (is the order in which it processes it twice consistent?). I guess what I'm saying is that I'd love to enable it by default so that annotations always appear (#1142) but I need to be a bit more confident that it won't bork 3D representations so I guess I'm asking is there something I haven't considered?

```
import ifcopenshell
import ifcopenshell.geom

f = ifcopenshell.open("/home/dion/Downloads/JudgesReception.ifc")
element = f.by_id(653)
body_3d = f.by_id(644)
axis_2d = f.by_id(633)

# *****************************************************************************

# INCLUDE_CURVES = True
# Method: Geometry iterator
# Element: Has 3D body + 2D axis
# Result: First 2D axis is created, then 3D body is created

print("Scenario A")
settings = ifcopenshell.geom.settings()
settings.set(settings.INCLUDE_CURVES, True)
iterator = ifcopenshell.geom.iterator(settings, f, include=[element])
iterator.initialize()
while True:
    shape = iterator.get()
    print(shape.guid)
    print(shape.geometry.edges)
    print(shape.geometry.faces)
    if not iterator.next():
        break

# *****************************************************************************

# INCLUDE_CURVES = True
# Method: Create shape on element
# Element: Has 3D body + 2D axis
# Result: Only 2D axis is created

print("Scenario B")
settings = ifcopenshell.geom.settings()
settings.set(settings.INCLUDE_CURVES, True)
shape = ifcopenshell.geom.create_shape(settings, element)
print(shape.guid)
print(shape.geometry.edges)
print(shape.geometry.faces)

# *****************************************************************************

# INCLUDE_CURVES = False
# Method: Create shape on element
# Element: Has 3D body + 2D axis
# Result: 3D body is created

print("Scenario C")
settings = ifcopenshell.geom.settings()
shape = ifcopenshell.geom.create_shape(settings, element)
print(shape.guid)
print(shape.geometry.edges)
print(shape.geometry.faces)

# *****************************************************************************

# INCLUDE_CURVES = False
# Method: Create shape on 3D body
# Result: 3D body is created

print("Scenario D")
settings = ifcopenshell.geom.settings()
shape = ifcopenshell.geom.create_shape(settings, body_3d)
print(shape.edges)
print(shape.faces)

# *****************************************************************************

# INCLUDE_CURVES = False
# Method: Create shape on 2D axis
# Result: Fail

print("Scenario E")
settings = ifcopenshell.geom.settings()
try:
    shape = ifcopenshell.geom.create_shape(settings, axis_2d)
except:
    print("Fail")

# *****************************************************************************

# INCLUDE_CURVES = True
# Method: Create shape on 3D body
# Result: 3D body is created

print("Scenario F")
settings = ifcopenshell.geom.settings()
settings.set(settings.INCLUDE_CURVES, True)
shape = ifcopenshell.geom.create_shape(settings, body_3d)
print(shape.edges)
print(shape.faces)


# *****************************************************************************

# INCLUDE_CURVES = True
# Method: Create shape on 2D axis
# Result: 2D axis is created

print("Scenario G")
settings = ifcopenshell.geom.settings()
settings.set(settings.INCLUDE_CURVES, True)
shape = ifcopenshell.geom.create_shape(settings, axis_2d)
print(shape.edges)
print(shape.faces)
```

**Comment by Moult** [comment]
Using the file from #1142 I get the opposite behaviour for scenario A and B - why?

```
f = ifcopenshell.open("/home/dion/Downloads/anno.ifc")
element = f.by_id(313)
body_3d = f.by_id(261)
axis_2d = f.by_id(296)


# *****************************************************************************

# INCLUDE_CURVES = True
# Method: Geometry iterator
# Element: Has 3D body + 2D axis
# Old Result: First 2D axis is created, then 3D body is created
# New Result: First 3D body is created, then 2D axis is created

print("Scenario A")
settings = ifcopenshell.geom.settings()
settings.set(settings.INCLUDE_CURVES, True)
iterator = ifcopenshell.geom.iterator(settings, f, include=[element])
iterator.initialize()
while True:
    shape = iterator.get()
    print(shape.guid)
    print(shape.geometry.edges)
    print(shape.geometry.faces)
    if not iterator.next():
        break

# *****************************************************************************

# INCLUDE_CURVES = True
# Method: Create shape on element
# Element: Has 3D body + 2D axis
# Old Result: Only 2D axis is created
# New Result: Only 3D body is created

print("Scenario B")
settings = ifcopenshell.geom.settings()
settings.set(settings.INCLUDE_CURVES, True)
shape = ifcopenshell.geom.create_shape(settings, element)
print(shape.guid)
print(shape.geometry.edges)
print(shape.geometry.faces)
```

**Comment by aothms** [comment]
I guess the order is not something to rely on (it's likely based on order in the file). Rather look at the `context` attribute or reverse engineer the representation it was created from:

```
>>> import ifcopenshell, ifcopenshell.geom
>>> f = ifcopenshell.open("duplex.ifc")
>>> settings = ifcopenshell.geom.settings(INCLUDE_CURVES=True)
>>> iterator = ifcopenshell.geom.iterator(settings, f, include=f.by_type("IfcWall"))
>>> print(*[(ob.guid, ob.context, ob.geometry.id, f[int(ob.geometry.id.split('-')[0])]) for ob in ifcopenshell.geom.iterate(settings, f, include=f.by_type("IfcWall"))])
('2O2Fr$t4X7Zf8NOew3FK3E', 'Body', '5262', #5262=IfcShapeRepresentation(#3,'Body','SweptSolid',(#5864))) ('0dxE1Sy6nDqfpDb5vIMNiA', 'Axis', '5266', #5266=IfcShapeRepresentation(#3,'Axis','Curve2D',(#18121))) ('2O2Fr$t4X7Zf8NOew3FK9N', 'Axis', '5299-material-1576', #5299=IfcShapeRepresentation(#3,'Axis','Curve2D',(#18345))) ...
```

With multiple curve-like representation all should be included (??)

For create_shape() the behaviour is indeed a bit different, but I would really use the optional third IfcRepresentation attribute if you need that kind of control.

**Comment by Moult** [comment]
I've just come across a scenario where using the geometry iterator with INCLUDE_CURVES=False, where doors have both a Model/Body/MODEL_VIEW Tessellation, and a Model/Footprint/MODEL_VIEW MappedRepresentation GeometricSet of IndexedPolyCurves... and the geometry iterator will run twice on the same element - first it will create the Tessellation, and _then_ it will create the Footprint. (The end result is that the user by default sees curves)

Is this desired behaviour? I would've thought that with INCLUDE_CURVES=False the footprint would not be processed?

**Comment by Moult** [comment]
Here is some minimal code which replicates the situation I'm coming across in the previous comment.

As you can see, INCLUDE_CURVES is explicitly set to false. Interestingly, on Linux, I get the desired behaviour - i.e., I only get a single Body representation. However, on Windows, I get _two_ results: once for the Body, and another for the Footprint.

Spooky. Thoughts?

[door-include-curves.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/6136287/door-include-curves.ifc.txt)

```
import ifcopenshell
import ifcopenshell.geom

f = ifcopenshell.open("/home/dion/door-include-curves.ifc")
element = f.by_id(160)

settings = ifcopenshell.geom.settings()
settings.set(settings.INCLUDE_CURVES, False)
iterator = ifcopenshell.geom.iterator(settings, f, include=[element])
iterator.initialize()
while True:
    shape = iterator.get()
    print(shape.guid)
    print(shape.geometry.edges)
    print(shape.geometry.faces)
    if not iterator.next():
        break
```

**Comment by plaes** [comment]
Would it make sense to implement these examples in the initial comment as part of testsuite?

**Comment by aothms** [comment]
Ok, I've figured it out. I think. It's probably not that there is a difference in windows/linux, it's just that very recently only I added support for geometric sets of curves https://github.com/IfcOpenShell/IfcOpenShell/commit/45f3e677dfc3a78566b994d068aa1183442526ef

But being wrapped in a geometric set, the "geometry type mapping function" shape_type() isn't sufficient anymore to exclude the geometry from processing. So this has to be handled specifically in the conversion function of geometric set. A bit annoying is that it results in a logged error now because the conversion function "fails".

@plaes definitely! If you or someone else is willing to setup a unit test structure I think it's a great way to structure this and automatically see it getting resolved.

(There is this strange https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/test/tests.py now too btw that is actually being ran on Travis, but should be turned into proper unit tests at some point).
